client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -361,11 +361,19 @@ public void cancelled() {
     private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
         Objects.requireNonNull(requestHeaders, "request headers must not be null");
         for (Header defaultHeader : defaultHeaders) {
-            httpRequest.setHeader(defaultHeader);
+            httpRequest.addHeader(defaultHeader);
         }
-        for (Header requestHeader : requestHeaders) {
-            Objects.requireNonNull(requestHeader, "request header must not be null");
-            httpRequest.setHeader(requestHeader);
+        if (requestHeaders.length != 0) {
+            final Set<String> uniqueNames = new HashSet<>();
+            for (Header requestHeader : requestHeaders) {
+                Objects.requireNonNull(requestHeader, "request header must not be null");
+                // overwrite any default header
+                if (uniqueNames.add(requestHeader.getName())) {
+                    httpRequest.setHeader(requestHeader);
+                } else {
+                    httpRequest.addHeader(requestHeader);
```

Comment:
I see why it is needed why. but I am afraid I will wonder again every day I look at this method :) Can we make the choice dependent on whether the header appears among the default ones too rather than using a set for the request headers? How does that sound?